File: BJ - 2167(2차원 배열의 합).java
	import java.util.*;
	import java.io.*;
	
	public class Main {
		static int sx,sy,ex,ey,N,M,K,ans;
		static int arr[][],sum_arr[][];
		public static void main(String[] args) throws IOException{
			BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
			BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
			StringTokenizer st;
			st = new StringTokenizer(br.readLine());
			N= Integer.parseInt(st.nextToken());
			M= Integer.parseInt(st.nextToken());
			arr = new int [N][M];
			sum_arr = new int [N+1][M+1];
			
			for(int i =0; i <N; i ++) {
				st = new StringTokenizer(br.readLine());
				for(int j =0; j <M; j ++) {
					arr[i][j] = Integer.parseInt(st.nextToken());
					sum_arr[i+1][j+1] = sum_arr[i][j+1] + sum_arr[i+1][j] - sum_arr[i][j] + arr[i][j];
				}
			}
		K = Integer.parseInt(br.readLine());
		for(int i =0; i <K; i ++) {
			ans = 0;
			st = new StringTokenizer(br.readLine());
			sx = Integer.parseInt(st.nextToken());
			sy = Integer.parseInt(st.nextToken());
			ex = Integer.parseInt(st.nextToken());
			ey = Integer.parseInt(st.nextToken());
			ans = sum_arr[ex][ey] -sum_arr[ex][sy-1] - sum_arr[sx-1][ey] + sum_arr[sx-1][sy-1];
			bw.write(ans + "\n");
		}
		bw.flush();
		br.close();
		bw.close();
	}
	}

File: BJ - 2167(2차원 배열의 합).py
import sys
input = sys.stdin.readline
N, M = map(int, input().split())
arr = [[0] * M for i in range(N)]
sum_arr = [[0] * (M+1) for i in range(N+1)]
for i in range(N):
    arr[i] = list(map(int, input().split()))
    for j in range(M):
        sum_arr[i+1][j+1] += arr[i][j] + sum_arr[i+1][j]+ sum_arr[i][j+1] - sum_arr[i][j]

# print(sum_arr)
K = int(input())
for i in range(K):
    sx,sy,ex,ey = map(int, input().split())
    # print(sum_arr[ex][ey],sum_arr[sx - 1][ey], sum_arr[ex][sy - 1] ,sum_arr[sx - 1][sy - 1])
    print(sum_arr[ex][ey] - sum_arr[sx-1][ey] - sum_arr[ex][sy-1] + sum_arr[sx-1][sy-1])

# 4 4
# 1 2 3 4
# 5 6 7 8
# 9 10 11 12
# 13 14 15 16
# 1
# 2 2 3 4